scipy/ndimage/src/_rank_filter_1d.cpp

```diff
@@ -0,0 +1,310 @@
+      maxSortDown(data, m, -1);
+    }
+  } else if (p < 0) // new item is in maxheap
+  {
+    if (v < old) {
+      maxSortDown(data, m, p);
+      return;
+    }
+    if (maxSortUp(data, m, p) && mmCmpExch(data, m, 1, 0)) {
+      minSortDown(data, m, 1);
+    }
+  } else // new item is at rank
+  {
+    if (maxSortUp(data, m, -1)) {
+      maxSortDown(data, m, -1);
+    }
+    if (minSortUp(data, m, 1)) {
+      minSortDown(data, m, 1);
+    }
+  }
+}
+
+template <typename T>
+void _rank_filter(T *in_arr, int rank, int arr_len, int win_len, T *out_arr,
+                  int mode, T cval, int origin) {
+  int i, arr_len_thresh, lim = (win_len - 1) / 2 - origin;
+  int lim2 = arr_len - lim;
+  Mediator *m = MediatorNew(win_len, rank);
+  T *data = new T[win_len];
+
+  switch (mode) {
+  case REFLECT:
+    for (i = win_len - lim - 1; i > -1; i--) {
+      MediatorInsert(data, m, in_arr[i]);
+    }
+    break;
+  case CONSTANT:
+    for (i = win_len - lim; i > 0; i--) {
+      MediatorInsert(data, m, cval);
+    }
+    break;
+  case NEAREST:
+    for (i = win_len - lim; i > 0; i--) {
+      MediatorInsert(data, m, in_arr[0]);
+    }
+    break;
+  case MIRROR:
+    for (i = win_len - lim; i > 0; i--) {
+      MediatorInsert(data, m, in_arr[i]);
+    }
+    break;
+  case WRAP:
+    for (i = arr_len - lim - 1 - 2 * origin; i < arr_len; i++) {
+      MediatorInsert(data, m, in_arr[i]);
+    }
+    break;
+  }
+
+  for (i = 0; i < lim; i++) {
+    MediatorInsert(data, m, in_arr[i]);
+  }
+  for (i = lim; i < arr_len; i++) {
+    MediatorInsert(data, m, in_arr[i]);
+    out_arr[i - lim] = data[m->heap[0]];
+  }
+  switch (mode) {
+  case REFLECT:
+    arr_len_thresh = arr_len - 1;
+    for (i = 0; i < lim; i++) {
+      MediatorInsert(data, m, in_arr[arr_len_thresh - i]);
+      out_arr[lim2 + i] = data[m->heap[0]];
+    }
+    break;
+  case CONSTANT:
+    for (i = 0; i < lim; i++) {
+      MediatorInsert(data, m, cval);
+      out_arr[lim2 + i] = data[m->heap[0]];
+    }
+    break;
+  case NEAREST:
+    arr_len_thresh = arr_len - 1;
+    for (i = 0; i < lim; i++) {
+      MediatorInsert(data, m, in_arr[arr_len_thresh]);
+      out_arr[lim2 + i] = data[m->heap[0]];
+    }
+    break;
+  case MIRROR:
+    arr_len_thresh = arr_len - 2;
+    for (i = 0; i < lim + 1; i++) {
+      MediatorInsert(data, m, in_arr[arr_len_thresh - i]);
+      out_arr[lim2 + i] = data[m->heap[0]];
+    }
+    break;
+  case WRAP:
+    for (i = 0; i < win_len; i++) {
+      MediatorInsert(data, m, in_arr[i]);
+      out_arr[lim2 + i] = data[m->heap[0]];
+    }
+    break;
+  }
+
+  m->heap -= rank;
+  delete[] m->heap;
+  m->heap = nullptr;
+  delete[] m->pos;
+  m->pos = nullptr;
+  delete m;
+  m = nullptr;
+  delete[] data;
+  data = nullptr;
+}
+
+// Python wrapper for rank_filter
+static PyObject *rank_filter(PyObject *self, PyObject *args) {
+  PyObject *in_arr_obj, *out_arr_obj, *cval_obj;
+  int rank, arr_len, win_len, mode, origin, type;
```

Comment:
@ev-br, I suspect that this int is in fact int16 in my system. Is it the proper way to write it? I am concerned that it might not work properly for arrays longer than 2^16.

Reply:
There really isn't a system (that we really support) where C's `int` is `int16_t`. Maybe some weird embedded platform, but I really don't expect much of the rest of scipy would work there.

Nonetheless, we do use `npy_intp` for things that might be array-length sizes. Check the rest of the `ndimage` codebase for examples. Some of these variables (like `mode` and `type`) do use use `int` in the rest of `ndimage`.